### ISSUE TITLE
Expose connection settings of communicator to template

### ISSUE THREAD

**Issue by timurb** [hashicorp/packer#4701]
It would be nice to have the variable with the IP of the host inside template.

My use case for this will be to run InSpec tests from local shell provisioner.
Something like:
```
{
  "type": "shell-local",
  "command": "inspec exec spec/ --sudo --sudo-password=packer --target=ssh://packer:packer@{{ssh_host}}"
}
```
I understand that there could be issues with bastion host and other SSH connection parameters but a template developer already knows these when writing the template.
The only dynamic value that I can't know in advance is the IP address of the host to connect.

I understand that I can workaround that by installing InSpec to inside the VM but I'd like to have an option to avoid that.
Another use case could be described in https://github.com/mitchellh/packer/issues/4173

This is similar to https://github.com/mitchellh/packer/issues/3889 but relates to templates rather than to provisioners.

**Comment by rickard-von-essen** [comment]
I think we have argued over this enough in #3331. @mwhooker close?

**Comment by timurb** [comment]
#3331 is a different case, it is about passing IP to provisioner.

My issue is a different one and is about passing IP to template. See `{{ssh_host}}` call in the example above.

**Comment by rickard-von-essen** [comment]
It's the same thing as described in the original post of #3331.

**Comment by derekmurawsky** [comment]
This was closed as "the same thing as described in... #3331. It is not. #3331 was closed because ansible was the only thing that needed this. This is a clear example of something else that needs it. 
I would submit that there are any number of reasons that having the IP readily available is necessary. Security scanners, serverspec tests, and any number of other tests that should happen before the AMI is bundled and published. 
Yes, there are other ways of doing so, bu they involve tailing logs and other such things. In our instance, we have the AWS CLI app, but no way to get the instance ID without hacking logs as well. 
Please reconsider your position on making this available somewhere. 

Perhaps it makes sense to make them available from specific builders instead of everywhere? For example the AWS builder could include an instance ID var and an internal IP var?

Bottom line is, I think this is important to cleanly extend the use of packer.

**Comment by kiwivogel** [comment]
Agreed. The alternative would be an inspec provisioner (would be even better).

**Comment by hogarthj** [comment]
Note that that the ansible provisioner, over winrm, is extremely fragile and because it bypasses the tested and expected route that ansible takes to reach windows (ie it doesn't use the ansible winrm connection code) it steps out of the tested scenarios for ansible.

Exposing the IP so that ansible-playbook can be run directly with shell-local makes so much more sense for maintainability and to avoid weird edge cases.

**Comment by SwampDragons** [comment]
I've been persuaded that this is necessary, and I want to put some time into figuring out how to implement this _properly_ across the various plugins.  It needs to be a core change, though, so it'll take some time and architectural refactoring.  I'm going to close this ticket and link it to the meta-ticket that'll handle that architectural refactor -- see #7079 above.
